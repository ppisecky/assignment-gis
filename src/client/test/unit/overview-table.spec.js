import { shallowMount, mount, createLocalVue } from '@vue/test-utils'
import OverviewTable from '../../components/overview-table'
import ElementUI from 'element-ui';
import { regions } from './__mocks__/mockedData'
import { randomRegionLevel, arrRandIndex } from './helpers'

const localVue = createLocalVue();
localVue.use(ElementUI);

const shallowMountOverviewTable = (settings = {}) => shallowMount(OverviewTable, { ...{ localVue }, ...settings })

describe('region overview table', () => {
    it('mounts properly', () => {
        // when
        const testSubject = shallowMountOverviewTable()

        // then
        expect(testSubject.isVueInstance()).toBeTruthy()
    })

    it('displays warning without data', () => {
        // when
        const testSubject = shallowMountOverviewTable()

        // then
        expect(testSubject.props().overview).toBe(undefined)
        expect(testSubject.props().levelHasCapitals).toBe(undefined)
        expect(testSubject.contains('div.warning')).toBe(true)
    })

    it('displays overview level in title', () => {
        // given
        const regionLevel = randomRegionLevel()
        const props = {
            overview: {
                level: regionLevel,
                data: []
            }
        }

        // when
        const testSubject = shallowMountOverviewTable({
            propsData: props
        })

        // then
        expect(testSubject.props().overview.level).toBe(regionLevel)
        expect(testSubject.contains('h2')).toBe(true)
        expect(testSubject.find('h2').text().includes(regionLevel)).toBe(true)
    })

    it('displays number of regions in title', () => {
        // given
        const props = {
            overview: {
                level: 'foo',
                data: regions
            }
        }

        // when
        const testSubject = shallowMountOverviewTable({
            propsData: props
        })

        // then
        expect(testSubject.contains('h2')).toBe(true)
        expect(testSubject.find('h2 small').text().includes(regions.length)).toBe(true)
    })

    it('renders table with proper number of columns', () => {
        // given
        const props = {
            overview: {
                level: 'regions',
                data: regions
            },
            levelHasCapitals: true
        }

        // when
        const testSubject = mount(OverviewTable, {
            localVue,
            propsData: props
        })

        // then
        expect(testSubject.contains('table')).toBe(true)
        expect(testSubject.findAll('div.el-table__header-wrapper table th')).toHaveLength(5)
        testSubject.setProps({ levelHasCapitals: false })
        expect(testSubject.findAll('div.el-table__header-wrapper table th')).toHaveLength(4)
    })

    it('renders table with proper number of rows', () => {
        // given
        const props = {
            overview: {
                level: 'regions',
                data: regions
            },
            levelHasCapitals: true
        }

        // when
        const testSubject = mount(OverviewTable, {
            localVue,
            propsData: props
        })

        // then
        expect(testSubject.findAll('#overview-table div.el-table__body-wrapper > table > tbody > tr').length).toBe(regions.length)
    })

    it('emits data on region select', () => {
        // given 
        const props = {
            overview: {
                level: 'regions',
                data: regions
            },
            levelHasCapitals: true
        }

        // when
        const testSubject = mount(OverviewTable, {
            localVue,
            propsData: props
        })
        const displayedRows = testSubject.findAll('#overview-table div.el-table__body-wrapper > table > tbody > tr')
        const rowIdx = arrRandIndex(displayedRows)
        displayedRows.at(rowIdx).trigger('click')

        // then
        expect(testSubject.emitted().areaSelected).toBeTruthy()
        expect(testSubject.emitted().areaSelected).toHaveLength(1)
        expect(testSubject.emitted().areaSelected[0]).toEqual([props.overview.data[rowIdx]])
    })

    it('renders correctly', () => {
        // given
        const props = {
            overview: {
                level: 'regions',
                data: regions
            },
            levelHasCapitals: true
        }

        // when
        const freshLocalVue = createLocalVue();
        freshLocalVue.use(ElementUI);
        const testSubject = mount(OverviewTable, {
            localVue: freshLocalVue,
            propsData: props
        })

        // then
        expect(testSubject).toMatchSnapshot();
    })
})