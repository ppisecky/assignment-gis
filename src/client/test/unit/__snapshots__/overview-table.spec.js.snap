// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`region overview table renders correctly 1`] = `
<div id="overview-table">
  <div>
    <h2>
      Overview level: regions
      <br> <small>4 results</small></h2>
    <div class="el-table el-table--fit el-table--fluid-height el-table--scrollable-x el-table--enable-row-transition">
      <div class="hidden-columns">
        <div></div>
        <div></div>
        <div></div>
        <div></div>
        <div></div>
      </div>
      <div class="el-table__header-wrapper">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 630px;">
          <colgroup>
            <col name="el-table_4_column_16" width="150">
            <col name="el-table_4_column_17" width="140">
            <col name="el-table_4_column_18" width="130">
            <col name="el-table_4_column_19" width="130">
            <col name="el-table_4_column_20" width="80">
          </colgroup>
          <thead class="">
            <tr class="">
              <th colspan="1" rowspan="1" class="el-table_4_column_16     is-hidden is-leaf is-sortable">
                <div class="cell">Name<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_4_column_17     is-leaf is-sortable">
                <div class="cell">Tourist places<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_4_column_18     is-leaf is-sortable">
                <div class="cell">Squares<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_4_column_19     is-leaf is-sortable">
                <div class="cell">Area<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
              </th>
              <th colspan="1" rowspan="1" class="el-table_4_column_20     is-leaf">
                <div class="cell">Capital</div>
              </th>
            </tr>
          </thead>
        </table>
      </div>
      <div class="el-table__body-wrapper is-scrolling-left" style="max-height: 162px;">
        <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 630px;">
          <colgroup>
            <col name="el-table_4_column_16" width="150">
            <col name="el-table_4_column_17" width="140">
            <col name="el-table_4_column_18" width="130">
            <col name="el-table_4_column_19" width="130">
            <col name="el-table_4_column_20" width="80">
          </colgroup>
          <tbody>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_4_column_16   is-hidden">
                <div class="cell">Treviso</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_17  ">
                <div class="cell">330</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_18  ">
                <div class="cell">22</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_19  ">
                <div class="cell">2477974475.61004</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_20  ">
                <div class="cell">Treviso</div>
              </td>
            </tr>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_4_column_16   is-hidden">
                <div class="cell">Vicenza</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_17  ">
                <div class="cell">629</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_18  ">
                <div class="cell">9</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_19  ">
                <div class="cell">2723437494.8171</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_20  ">
                <div class="cell">Vicenza</div>
              </td>
            </tr>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_4_column_16   is-hidden">
                <div class="cell">Venezia</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_17  ">
                <div class="cell">855</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_18  ">
                <div class="cell">11</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_19  ">
                <div class="cell">2476799680.41258</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_20  ">
                <div class="cell"></div>
              </td>
            </tr>
            <tr class="el-table__row">
              <td rowspan="1" colspan="1" class="el-table_4_column_16   is-hidden">
                <div class="cell">Padova</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_17  ">
                <div class="cell">439</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_18  ">
                <div class="cell">13</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_19  ">
                <div class="cell">2140647028.72377</div>
              </td>
              <td rowspan="1" colspan="1" class="el-table_4_column_20  ">
                <div class="cell">Padova</div>
              </td>
            </tr>
            <!---->
          </tbody>
        </table>
        <!---->
        <!---->
      </div>
      <!---->
      <div class="el-table__fixed" style="bottom: 0px; width: 150px;">
        <div class="el-table__fixed-header-wrapper">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__header" style="width: 630px;">
            <colgroup>
              <col name="el-table_4_column_16" width="150">
              <col name="el-table_4_column_17" width="140">
              <col name="el-table_4_column_18" width="130">
              <col name="el-table_4_column_19" width="130">
              <col name="el-table_4_column_20" width="80">
            </colgroup>
            <thead class="">
              <tr class="">
                <th colspan="1" rowspan="1" class="el-table_4_column_16     is-leaf is-sortable">
                  <div class="cell">Name<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_4_column_17     is-hidden is-leaf is-sortable">
                  <div class="cell">Tourist places<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_4_column_18     is-hidden is-leaf is-sortable">
                  <div class="cell">Squares<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_4_column_19     is-hidden is-leaf is-sortable">
                  <div class="cell">Area<span class="caret-wrapper"><i class="sort-caret ascending"></i><i class="sort-caret descending"></i></span></div>
                </th>
                <th colspan="1" rowspan="1" class="el-table_4_column_20     is-hidden is-leaf">
                  <div class="cell">Capital</div>
                </th>
              </tr>
            </thead>
          </table>
        </div>
        <div class="el-table__fixed-body-wrapper" style="top: 44px; max-height: 162px;">
          <table cellspacing="0" cellpadding="0" border="0" class="el-table__body" style="width: 630px;">
            <colgroup>
              <col name="el-table_4_column_16" width="150">
              <col name="el-table_4_column_17" width="140">
              <col name="el-table_4_column_18" width="130">
              <col name="el-table_4_column_19" width="130">
              <col name="el-table_4_column_20" width="80">
            </colgroup>
            <tbody>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_4_column_16  ">
                  <div class="cell">Treviso</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_17   is-hidden">
                  <div class="cell">330</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_18   is-hidden">
                  <div class="cell">22</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_19   is-hidden">
                  <div class="cell">2477974475.61004</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_20   is-hidden">
                  <div class="cell">Treviso</div>
                </td>
              </tr>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_4_column_16  ">
                  <div class="cell">Vicenza</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_17   is-hidden">
                  <div class="cell">629</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_18   is-hidden">
                  <div class="cell">9</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_19   is-hidden">
                  <div class="cell">2723437494.8171</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_20   is-hidden">
                  <div class="cell">Vicenza</div>
                </td>
              </tr>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_4_column_16  ">
                  <div class="cell">Venezia</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_17   is-hidden">
                  <div class="cell">855</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_18   is-hidden">
                  <div class="cell">11</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_19   is-hidden">
                  <div class="cell">2476799680.41258</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_20   is-hidden">
                  <div class="cell"></div>
                </td>
              </tr>
              <tr class="el-table__row">
                <td rowspan="1" colspan="1" class="el-table_4_column_16  ">
                  <div class="cell">Padova</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_17   is-hidden">
                  <div class="cell">439</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_18   is-hidden">
                  <div class="cell">13</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_19   is-hidden">
                  <div class="cell">2140647028.72377</div>
                </td>
                <td rowspan="1" colspan="1" class="el-table_4_column_20   is-hidden">
                  <div class="cell">Padova</div>
                </td>
              </tr>
              <!---->
            </tbody>
          </table>
          <!---->
        </div>
        <!---->
      </div>
      <!---->
      <!---->
      <div class="el-table__column-resize-proxy" style="display: none;"></div>
    </div>
  </div>
</div>
`;
